daemon/k8s_watcher.go

```diff
@@ -1429,6 +1432,36 @@ func (d *Daemon) deleteCiliumNetworkPolicyV2(cnp *cilium_v2.CiliumNetworkPolicy)
 	}
 }
 
+func updateRuleAnnotations(newRuleCpy *cilium_v2.CiliumNetworkPolicy) error {
+
+	// Create DeepCopy, as newRuleCpy is a pointer and may be modified elsewhere
+	// outside of this function, and we want to ensure that the annotations
+	// which are extracted are not modified.
+	ruleCopyForAnnotations := newRuleCpy.DeepCopy()
+
+	// Only update annotations for node on which this agent is running.
+	nodeName := node.GetName()
+	cnpns := ruleCopyForAnnotations.GetPolicyStatus(nodeName)
+	cnpns.Annotations = ruleCopyForAnnotations.Annotations
+
+	// Do not update annotations if policy rule failed to be installed previously.
```

Comment:
I'm probably missing some context, but what does the policy rule installations has to do with the annotations?

Reply:
https://github.com/cilium/cilium/issues/4248

Reply:
If users are expecting the annotation to be the source of truth when a policy is being enforced, why then, when an annotation is changed aren't we triggering a policy regeneration by deleting and adding the CNP?

Reply:
The fix in this PR is to update the annotations when the spec / specs are not changed in the policy. If the annotation is changed by the user, we still update the annotations but do not trigger policy updates.

Reply:
>The fix in this PR is to update the annotations when the spec / specs are not changed in the policy. If the annotation is changed by the user, we still update the annotations but do not trigger policy updates.

Then, the annotation can't be used as a source of truth if the policy is being enforced, which AFAIK is what #4248 is all about.

Reply:
Why? The policy is already enforced if the spec / specs haven't changed and the Status for that CNP on that node is OK, isn't it? I don't see why we have to delete / re-add the CNP and re-trigger policy updates, which is extremely costly, if only the annotations for the policy, which have no bearing on the datapath itself, has changed. 

This change is saying that if the policy is already enforced, and has no datapath-related changes, then update the annotations directly.

Reply:
OK I got it.